cmd/node_builder.go

```diff
@@ -0,0 +1,130 @@
+	ParseAndPrintFlags()
+
+	// Initialize performs all the initialization needed at the very start of a node
+	Initialize() NodeBuilder
+
+	// PrintBuildVersionDetails prints the node software build version
+	PrintBuildVersionDetails()
+
+	// EnqueueNetworkInit enqueues the default network component
+	EnqueueNetworkInit()
+
+	// EnqueueMetricsServerInit enqueues the metrics component
+	EnqueueMetricsServerInit()
+
+	// Enqueues the Tracer component
+	EnqueueTracer()
+
+	// Module enables setting up dependencies of the engine with the builder context
+	Module(name string, f func(builder NodeBuilder, node *NodeConfig) error) NodeBuilder
+
+	// Component adds a new component to the node that conforms to the ReadyDone
+	// interface.
+	//
+	// When the node is run, this component will be started with `Ready`. When the
+	// node is stopped, we will wait for the component to exit gracefully with
+	// `Done`.
+	Component(name string, f func(builder NodeBuilder, node *NodeConfig) (module.ReadyDoneAware, error)) NodeBuilder
+
+	// MustNot asserts that the given error must not occur.
+	// If the error is nil, returns a nil log event (which acts as a no-op).
+	// If the error is not nil, returns a fatal log event containing the error.
+	MustNot(err error) *zerolog.Event
+
+	// Run initiates all common components (logger, database, protocol state etc.)
+	// then starts each component. It also sets up a channel to gracefully shut
+	// down each component if a SIGINT is received.
+	Run()
+
+	// PreInit registers a new PreInit function.
+	// PreInit functions run before the protocol state is initialized or any other modules or components are initialized
+	PreInit(f func(builder NodeBuilder, node *NodeConfig)) NodeBuilder
+
+	// PostInit registers a new PreInit function.
+	// PostInit functions run after the protocol state has been initialized but before any other modules or components
+	// are initialized
+	PostInit(f func(builder NodeBuilder, node *NodeConfig)) NodeBuilder
+
+	// RegisterBadgerMetrics registers all badger related metrics
+	RegisterBadgerMetrics()
+}
+
+// BaseConfig is the general config for the NodeBuilder and the command line params
+type BaseConfig struct {
+	nodeIDHex             string
+	bindAddr              string
+	NodeRole              string
+	timeout               time.Duration
+	datadir               string
+	level                 string
+	metricsPort           uint
+	BootstrapDir          string
+	peerUpdateInterval    time.Duration
+	unicastMessageTimeout time.Duration
+	profilerEnabled       bool
+	profilerDir           string
+	profilerInterval      time.Duration
+	profilerDuration      time.Duration
+	tracerEnabled         bool
+}
+
+// NodeConfig contains all the derived parameters such the NodeID, private keys etc. and initialized instances of
+// structs such as DB, Network etc. The NodeConfig is composed of the BaseConfig and is updated in the
+// NodeBuilder functions as a node is bootstrapped.
+type NodeConfig struct {
```

Comment:
@huitseeker @smnzhu @jordanschalm Added this new struct to hold all the data for the node. The `NodeBuilder` functions modifies this struct through its methods.